File: Src/Icm.Core/Functions/TotalOrder/BaseTotalOrder.vb
Namespace Icm

    Public MustInherit Class BaseTotalOrder(Of T As IComparable(Of T))
        Implements ITotalOrder(Of T)

        Public MustOverride Function Least() As T Implements ITotalOrder(Of T).Least

        Public MustOverride Function Long2T(ByVal d As Long) As T Implements ITotalOrder(Of T).Long2T

        Public MustOverride Function Greatest() As T Implements ITotalOrder(Of T).Greatest

        Public MustOverride Function T2Long(ByVal t As T) As Long Implements ITotalOrder(Of T).T2Long

        Public Overridable Function Compare(ByVal x As T, ByVal y As T) As Integer Implements System.Collections.Generic.IComparer(Of T).Compare
            Return T2Long(x).CompareTo(T2Long(y))
        End Function

        Public Overridable Function [Next](ByVal t As T) As T Implements ITotalOrder(Of T).Next
            Dim longVal As Long = T2Long(t)
            Dim longGST As Long = T2Long(Greatest)
            Dim result As T

            Do
                If longVal >= longGST Then
                    Throw New ArgumentOutOfRangeException("t")
                End If
                longVal += 1
                result = Long2T(longVal)
            Loop Until t.CompareTo(result) < 0

            Return result
        End Function

        Public Overridable Function Previous(ByVal t As T) As T Implements ITotalOrder(Of T).Previous
            Dim longVal As Long = T2Long(t)
            Dim longLST As Long = T2Long(Least)
            Dim result As T

            Do
                If longVal <= longLST Then
                    Throw New ArgumentOutOfRangeException("t")
                End If
                longVal -= 1
                result = Long2T(longVal)
            Loop Until t.CompareTo(result) > 0

            Return result
        End Function

    End Class
End Namespace

File: Src/Icm.Core/Functions/TotalOrder/BaseTotalOrder.cs
using System;

namespace Icm
{

    public abstract class BaseTotalOrder<T> : ITotalOrder<T> where T : IComparable<T>
    {

        public abstract T Least();

        public abstract T Long2T(long d);

        public abstract T Greatest();

        public abstract long T2Long(T t);

        public virtual int Compare(T x, T y)
        {
            return T2Long(x).CompareTo(T2Long(y));
        }

        public virtual T Next(T t)
        {
            long longVal = T2Long(t);
            long longGst = T2Long(Greatest());
            T result;

            do
            {
                if (longVal >= longGst)
                {
                    throw new ArgumentOutOfRangeException(nameof(t));
                }
                longVal += 1;
                result = Long2T(longVal);
            } while (!(t.CompareTo(result) < 0));

            return result;
        }

        public virtual T Previous(T t)
        {
            long longVal = T2Long(t);
            long longLst = T2Long(Least());
            T result;

            do
            {
                if (longVal <= longLst)
                {
                    throw new ArgumentOutOfRangeException(nameof(t));
                }
                
                longVal -= 1;
                result = Long2T(longVal);
            } while (!(t.CompareTo(result) > 0));

            return result;
        }

    }
}
